### PR TITLE
fix: update rpc on continuous sync

### LINKED ISSUE #2705: Hive `ethereum/sync` test regression
### Describe the bug

The hive `ethereum/sync` test currently fails:
https://github.com/paradigmxyz/reth/actions/runs/4986338461/jobs/8927244103

This is probably due to not updating the `latest` block in RPC. We are not sent forkchoice updates or payloads in the sync test, so we do not set the canonical head, which is currently used to determine which block to return when given the `latest` argument.

### Steps to reproduce

Run hive:
```
./hive --client reth --sim "ethereum/sync"
```

### Node logs

_No response_

### Platform(s)

_No response_

### What version/commit are you on?

_No response_

### Code of Conduct

- [X] I agree to follow the Code of Conduct

### PR DESCRIPTION
**Pull request by Rjected**
The `latest` tag should always be consistent with the canonical head. Before we used the `ChainInfoTracker`, we would update the latest head in the `Finish` stage, which is not correct. When not using `--debug.continuous`, we correctly update the `ChainInfoTracker`, and latest head, on a successful forkchoice update. Because `--debug.continuous` is a nonstandard sync strategy that does not rely on forkchoice updates or other CL requests, we should update the `latest` block whenever the pipeline finishes, preserving the behavior that the `Finish` stage previously fulfilled.

This PR updates the canonical head when running with `--debug.continuous` with the stage progress, whenever the pipeline finishes.

Fixes #2705 

~~This is a super hacky solution to fixing the sync test, because it inserts continuous downloader logic into the consensus engine. Mainly submitting this to highlight that we need to update how we set the `latest` block.~~

We currently only ever set the `latest` block on engine API calls, ~~and conflate the latest block with the canonical head~~ (update: this is what we should be doing and consistent with geth). ~~We might want to go back to setting the latest block in the finish stage, or something similar.~~

### CODE REVIEW

File: crates/consensus/beacon/src/engine/mod.rs [hunk 1]
```diff
@@ -799,6 +799,32 @@ where
...
+                                Ok(Some(block)) => block,
+                                Ok(None) => {
+                                    return Some(Err(Error::Provider(
+                                        ProviderError::HeaderNotFound(max_block.into()),
+                                    )
+                                    .into()))
+                                }
+                                Err(error) => return Some(Err(error.into())),
+                            };
+                            let max_header = match self.blockchain.block_by_number(max_block) {
```
Comment (rkrasiuk) [inline crates/consensus/beacon/src/engine/mod.rs:815]: no need to look up the full block

File: crates/consensus/beacon/src/engine/mod.rs [hunk 2]
```diff
@@ -799,6 +799,32 @@ where
                             return Some(Ok(()))
                         }
 
+                        // update the canon chain if continuous is enabled
+                        if self.sync.run_pipeline_continuously() {
+                            let max_block = ctrl.progress().unwrap_or_default();
```
Comment (rkrasiuk) [inline crates/consensus/beacon/src/engine/mod.rs:804]: extracting best block lookup to a separate function will be cleaner in terms of error handlig

**Review by rkrasiuk — CHANGES_REQUESTED** [review]

File: crates/consensus/beacon/src/engine/mod.rs (hunk 1 above)
Comment (mattsse) [inline crates/consensus/beacon/src/engine/mod.rs:815]: This is only the case when the pipeline finished, right, because we already update this on FCU?

this is a bit hard to follow and i'd like to perhaps move this to a separate function if possible with some docs.

**Review by mattsse — CHANGES_REQUESTED** [review]
nice find!

File: crates/consensus/beacon/src/engine/mod.rs (hunk 1 above)
Comment (Rjected) [reply crates/consensus/beacon/src/engine/mod.rs]: right, now this is a separate method which only fetches the header

Comment (Rjected) [reply crates/consensus/beacon/src/engine/mod.rs]: Yes, although we only update this while handling the FCU if we already have the blocks contained in the FCU.

Added a separate method

File: crates/stages/src/pipeline/ctrl.rs [hunk 3]
```diff
@@ -32,4 +32,13 @@ impl ControlFlow {
     pub fn is_unwind(&self) -> bool {
         matches!(self, ControlFlow::Unwind { .. })
     }
+
+    /// Returns the pipeline progress, if the state is not `Unwind`.
+    pub fn progress(&self) -> Option<BlockNumber> {
```
Comment (mattsse) [inline crates/stages/src/pipeline/ctrl.rs:37]: I think we should consider renaming this and getting rid of the word "progress" because there is some ambiguity here. 

what does this really mean?
number of blocks since then last invokation? or current height after pipeline run?

but for now this is ok, because this follows the controlflow variant

cc @shekhirin

File: crates/consensus/beacon/src/engine/mod.rs [hunk 4]
```diff
@@ -866,6 +879,23 @@ where
...
+        block_number: BlockNumber,
+    ) -> Result<SealedHeader, reth_interfaces::Error> {
+        let max_hash = match self.blockchain.block_hash(block_number)? {
+            Some(block) => block,
+            None => return Err(Error::Provider(ProviderError::HeaderNotFound(block_number.into()))),
+        };
+        let max_header = match self.blockchain.header_by_number(block_number)? {
+            Some(block) => block,
+            None => return Err(Error::Provider(ProviderError::HeaderNotFound(block_number.into()))),
+        };
```
Comment (mattsse) [inline crates/consensus/beacon/src/engine/mod.rs:896]: this should be a provider function instead,

see #2769

**Review by mattsse — CHANGES_REQUESTED** [review]
I'd like to move the SealedHeader getter to a provider function

File: crates/stages/src/pipeline/ctrl.rs (hunk 3 above)
Comment (Rjected) [reply crates/stages/src/pipeline/ctrl.rs:37]: > or current height after pipeline run?

Ideally this is the correct definition, or at it's least what we need for continuous sync

File: crates/consensus/beacon/src/engine/mod.rs (hunk 4 above)
Comment (Rjected) [reply crates/consensus/beacon/src/engine/mod.rs]: Added in a separate PR, #2801

I can add a followup to remove this method and use the provider method

**Review by mattsse — DISMISSED** [review]
lgtm, can merge after rebase and replace the get_canonical_header function

File: crates/stages/src/pipeline/ctrl.rs (hunk 3 above)
Comment (shekhirin) [reply crates/stages/src/pipeline/ctrl.rs:37]: agree, I feel like progress has multiple definitions in the codebase and ideally it'd be nice to separate it into something like `checkpoint` (we use it to continue execution or unwind on next iteration), `block_height` (highest block number reached by a stage) and actual `progress` (stage-specific progress or block number)
